vllm/core/scheduler.py

```diff
@@ -56,24 +56,35 @@ class SchedulingBudget:
     max_num_seqs: int
     _request_ids_num_batched_tokens: Set[str] = field(default_factory=set)
     _request_ids_num_curr_seqs: Set[str] = field(default_factory=set)
+    # Number of cached tokens in the batch.
+    _num_cached_tokens: int = 0
+    # Number of actual non-cached tokens in the batch.
     _num_batched_tokens: int = 0
     _num_curr_seqs: int = 0
 
     def can_schedule(self, *, num_new_tokens: int, num_new_seqs: int):
-        assert num_new_tokens != 0
+        # We allow num_new_tokens to be 0 when the entire sequence has
+        # been cached.
+        assert num_new_tokens >= 0
```

Comment:
Actually I se all use cases of this function have a pre-condition `if num_nre_tokens==0 or can_schedule(num_new_tokens)`, so I guess this assertion still works?

Reply:
We actually have codes like `if all_new_tokens ==0 or not can_schedule(uncached_new_tokens` with the change now - so sometimes all new tokens count is not 0, but the uncached token count is 0.

Reply:
Ah I see. Then we could just improve the comment a bit

```
        # num_new_tokens may be 0 when the entire sequence has
        # been cached.
```

Reply:
btw what happen when an entire sequence is cached? Seems like we can always schedule it? But AFAIK our model runner doesn't support query_len=0 so it would be problematic if you send a sequence without input tokens. I'm wondering how this PR overcome this issue?

Reply:
Oh yeah, in this case, the number of cached tokens would be > 0, so the `token_chunk_size` would be > 0. It's just that this sequence doesn't contribute to budgeting anymore, but the new token count should always be >0. I can add an assert somewhere for this. 

But I do see in the main where we never return all blocks as cached when calculating computed blocks before, but I so far haven't seen any issue in my PR. Maybe I can validate it a bit more with some tests.